data/ooniprobe.1
.\" Man page generated from reStructuredText.
.
.TH "OONIPROBE" "1" "Feb 14, 2018" "2.3.0" "OONI: Open Observatory of Network Interference"
.SH NAME
ooniprobe \- an internet censorship measurement tool
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBooniprobe\fP [\fIoptions\fP] ([\fItest name\fP] | [\fIpath to nettest\fP\&.py])
.SH DESCRIPTION
.sp
\fBooniprobe\fP, is a tool for performing internet censorship
measurements. Our goal is to achieve a common data format and set of
methodologies for conducting censorship related research.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit.
.TP
.B \-n\fP,\fB  \-\-no\-collector
Disable writing to collector
.TP
.B \-N\fP,\fB  \-\-no\-njson
Disable writing to disk
.TP
.B \-g\fP,\fB  \-\-no\-geoip
Disable geoip lookup on start.
Warning: By using this option the IP address of the user executing ooniprobe is not excluded from the report.
.TP
.B \-s\fP,\fB  \-\-list
List the currently installed ooniprobe nettests
.TP
.B \-v\fP,\fB  \-\-verbose
Show more verbose information
.TP
.B \-w\fP,\fB  \-\-web\-ui
Start the web UI
.TP
.B \-z\fP,\fB  \-\-initialize
Initialize ooniprobe to begin running it
.TP
.BI \-o\fP,\fB  \-\-reportfile \ PATH_TO_FILE
Specify the report file name to write to.
.TP
.BI \-i\fP,\fB  \-\-testdeck \ PATH_TO_DECK
Specify as input a test deck: a yaml file containing the tests to run and their arguments.
.TP
.BI \-c\fP,\fB  \-\-collector \ COLLECTOR_ADDRESS
Specify the address of the collector for test results. In most cases a user
will prefer to specify a bouncer over this.
.TP
.BI \-b\fP,\fB  \-\-bouncer \ BOUNCER_ADDRESS
Specify the bouncer used to obtain the address of the collector and test helpers.
.TP
.BI \-l\fP,\fB  \-\-logfile \ PATH_TO_LOGFILE
Write to this logs to this filename.
.TP
.BI \-O\fP,\fB  \-\-pcapfile \ PATH_TO_PCAPFILE
Write a PCAP of the ooniprobe session to this filename.
.TP
.BI \-f\fP,\fB  \-\-configfile \ PATH_TO_CONFIG
Specify a path to the ooniprobe configuration file.
.TP
.B \-d\fP,\fB  \-\-datadir
Specify a path to the ooniprobe data directory.
.UNINDENT
.INDENT 0.0
.TP
.B \-a, –annotations key:value[,key2:value2]
Annotate the report with a key:value[, key:value] format.
.TP
.B \-P, –preferred\-backend onion|https|cloudfront
Set the preferred backend to use when submitting results and/or
communicating with test helpers. Can be either onion, https or cloudfront
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Display the ooniprobe version and exit.
.UNINDENT
.SH OONIPROBE
.SS Read this before running ooniprobe!
.sp
Running ooniprobe is a potentially risky activity. This greatly depends on the
jurisdiction in which you are in and which test you are running. It is
technically possible for a person observing your internet connection to be
aware of the fact that you are running ooniprobe. This means that if running
network measurement tests is something considered to be illegal in your country
then you could be spotted.
.sp
Furthermore, ooniprobe takes no precautions to protect the install target machine
from forensics analysis.  If the fact that you have installed or used ooni
probe is a liability for you, please be aware of this risk.
.SS What is this?
.sp
ooniprobe is the command line tool that volunteers and researches interested in
contributing data to the project should be running.
.sp
If you are interested in using ooniprobe from a graphical user interface
refer to \fBooniprobe\-agent\fP and see how to run that.
.sp
ooniprobe allows the user to select what test should be run and what backend
should be used for storing the test report and/or assisting them in the running
of the test.
.sp
ooniprobe tests are divided into two categories: \fBTraffic Manipulation\fP and
\fBContent Blocking\fP\&.
.sp
\fBTraffic Manipulation\fP tests aim to detect the presence of some sort of
tampering with the internet traffic between the probe and a remote test helper
backend. As such they usually require the selection of a oonib backend
component for running the test.
.sp
\fBContent Blocking\fP are aimed at enumerating the kind of content that is
blocked from the probes network point of view. As such they usually require to
have specified an input list for running the test.
.SH EXAMPLES
.sp
Run the web_connectivity test on \fI\%http://torproject.org\fP:
.INDENT 0.0
.INDENT 3.5
\fBooniprobe web_connectivity \-\-url http://torproject.org/\fP
.UNINDENT
.UNINDENT
.sp
Run the http_invalid_request_line test to detect middleboxes:
.INDENT 0.0
.INDENT 3.5
\fBooniprobe http_invalid_request_line\fP
.UNINDENT
.UNINDENT
.sp
Run the http_header_field_manipulation test to detect middleboxes:
.INDENT 0.0
.INDENT 3.5
\fBooniprobe http_header_field_manipulation\fP
.UNINDENT
.UNINDENT
.sp
List all the available tests:
.INDENT 0.0
.INDENT 3.5
\fBooniprobe \-s\fP
.UNINDENT
.UNINDENT
.sp
Start the web user interface:
.INDENT 0.0
.INDENT 3.5
\fBooniprobe \-w\fP
.UNINDENT
.UNINDENT
.SH AUTHOR
The Tor Project
.SH COPYRIGHT
2016, The Tor Project
.\" Generated by docutils manpage writer.
.


ChangeLog.md
## v1.1.0

In this new release of ooniprobe we have added a new command line tool for
listing the reports that have not been published to a collector and that allows
the probe operator to choose which ones they would like to upload.

We have also made some privacy improvements to the reports (we will sanitize
all things that may look like file paths) and added metadata associated with
the maxmind database being used by the probe operator.

Here is a more detailed list of what has been done:

* Annotate on disk which reports we have submitted and which ones we have not:
  https://trac.torproject.org/projects/tor/ticket/11860

* Add tool called oonireport for publishing unpublished ooniprobe reports to a
  collector: https://trac.torproject.org/projects/tor/ticket/11862

* Probe Report does not leak filepaths anymore:
  https://trac.torproject.org/projects/tor/ticket/12706

* Reports now include version information about the maxmind database being
  used: https://trac.torproject.org/projects/tor/ticket/12771

* We now do integrity checks on the ooniprobe.conf file so that we don't start
  the tool if the config file is missing some settings or is not consistent:
  https://trac.torproject.org/projects/tor/ticket/11983
  (thanks to Alejandro López (kudrom))

* Improvements have been made to the sniffer subsystem (thanks to Alejandro
  López (kudrom))

Minor bug fixes:

* Fix dns_spoof test (by kudrom)
  https://trac.torproject.org/projects/tor/ticket/12486

* ooni might not look at requiresTor:
  https://trac.torproject.org/projects/tor/ticket/11858

* ooni spits out gobs of tracebacks if Tor is not running and the OONI config
  says it will be:
  https://trac.torproject.org/projects/tor/ticket/11859

* The README for ooni-probe should mention the bugtracker and repository
  https://trac.torproject.org/projects/tor/ticket/11980

## v1.0.2

* Add ooniprobe manpage.

* Fix various security issues raised by the least authority audit.

* Add a test that checks for Tor bridge reachability.

* Record the IP address of the exit node being used in torified requests.

* Captive portal test now uses the ooni-probe test templates.

* Have better test naming consistency.

## v1.0.1

* Fix bugs in the traceroute test that lead to not all packets being collected.

* All values inside of http_requests test are now initialized inside of setUp.

* Fix a bug that lead to the input value of the report not being set in some
  circumstances.

* Add bridge_reachability test

## v1.0.0

* Add bouncer support for discovering test helpers and collectors

* Fix bug that lead to HTTP tests to stall

* Add support for connect_error and connection_lost_error error types

* Add support for additional Tor configuration keys

* Add disclaimer when running ooniprobe

## v0.1.0

Improvements to HTML/JS based user interface:

  * XSRF protection

  * user supplied input specification

Bugfixing and improvements to scheduler.

## v0.0.12

Implement JS/HTML based user interface.

Supports:

  * Starting and stopping of tests

  * Monitoring of test progress

## v0.0.11

* Parametrize task timeout and retry count

* Set the default collector via the command line option

* Add option to disable the default collector

* Add continuous integration with travis

## v0.0.10

### ooniprobe

* Fix bug that made HTTP based tests stall

* Update DNS Test example to not import the DNS Test template If you import the
	DNS Test template it will be considered a valid test case and command line
	argument parsing will not work as expected. see:
	#7795 for more details

* Fix major bug in DNS test template that prevented PTR lookups from working
	properly I was calling the queryUDP function with the arguments in the wrong
	order. Twisted, why you API no consistent?

* Add support for specifying the level of parallelism in tests (aka router
	melt mode)

* Do not swallow failures when a test instance fails to run fixes #7714

### scripts

* Add report archival script

Fix bug in TCP connect test that made it not properly log errors

* Refactor failure handling code in nettest Add function that traps all the
	supported failure and outputs the failure string representing it.

### documentation

* Add birdseye view of the ooniprobe architecture

* Add details on the current implementation status of ooni*

* Add draft ooniprobe API specification

* Add instructions for supervisord configuration and clean up README.md

## 0.0.9

### ooniprobe

* Set the default ASN to 0

* Make Beautiful soup a soft depedency

* Add support for sending the ASN number of the probe:
	the ASN number will get sent when creating a new report

* Add support for obtaining the probes IP address via getinfo address as per
	https://trac.torproject.org/projects/tor/ticket/7447

* Fix bug in ooniprobe test decks
	https://trac.torproject.org/projects/tor/ticket/7664

### oonib

* Use twisted fdesc when writing to files

* Add support for processing the ASN number of the probe

* Test reports shall follow the specification detailed inside of docs/reports.rst

* Add support for setting the tor binary path in oonib/config.py

### scripts

* Add a very simple example on how to securely parse the ooniprobe reports

### documentation

* Add documentation for the DNSSpoof test

* Add documentation for HTTPHeaderFieldManipulation

* Clean up writing_tests.rst

* Properly use the power of sphinx!

### Tests

* fixup Netalyzr third party plugin

## v0.0.8-alpha

### ooniprobe

* Allow test resolver file to have comments.

* Autostart Tor in default configuration.

* Add support for starting Tor via txtorcon.

* Make the sniffer not run in a separate thread, but use a non blocking fdesc.
	Do some refactoring of scapy testing, following Factory creational pattern
	and a pub-sub pattern for the readers and writers.

* Extend TrueHeaders to support calculation of difference between two HTTP headers respectful of
	capitalization


ooni/scripts/ooniprobe.py
#!/usr/bin/env python
import webbrowser
from multiprocessing import Process

from twisted.internet import task, defer

def ooniprobe(reactor):
    from ooni.ui.cli import runWithDaemonDirector, runWithDirector
    from ooni.ui.cli import setupGlobalOptions, initializeOoniprobe
    from ooni.settings import config

    global_options = setupGlobalOptions(logging=True, start_tor=True,
                                        check_incoherences=True)
    if global_options['info']:
        config.log_info()
        return defer.succeed(None)

    if global_options['queue']:
        return runWithDaemonDirector(global_options)

    if global_options['web-ui']:
        from ooni.settings import config
        from ooni.scripts.ooniprobe_agent import status_agent, start_agent
        if status_agent() != 0:
            p = Process(target=start_agent)
            p.start()
            p.join()
            print("Started ooniprobe-agent")
        webbrowser.open_new(config.web_ui_url)
        return defer.succeed(None)

    if global_options['initialize']:
        initializeOoniprobe(global_options)
        return defer.succeed(None)

    return runWithDirector(global_options)

def run():
    task.react(ooniprobe)

if __name__ == "__main__":
    run()
